packages/flutter/lib/src/material/text_selection_toolbar.dart

```diff
@@ -0,0 +1,721 @@
+    this.toolbarBuilder = _defaultToolbarBuilder,
+    required this.children,
+  }) : super(key: key);
+
+  /// The focal point above which the toolbar attempts to position itself.
+  ///
+  /// If there is not enough room above before reaching the top of the screen,
+  /// then the toolbar will position itself below [anchorBelow].
+  final Offset anchorAbove;
+
+  /// The focal point below which the toolbar attempts to position itself, if it
+  /// doesn't fit above [anchorAbove].
+  final Offset anchorBelow;
+
+  /// The children that will be displayed in the text selection toolbar.
+  ///
+  /// Typically these are buttons.
+  ///
+  /// See also:
+  ///   * [TextSelectionToolbarTextButton], which builds a default Material-
+  ///     style text selection toolbar text button.
+  final List<Widget> children;
+
+  /// Builds the toolbar container.
+  ///
+  /// Useful for customizing the high-level background of the toolbar. The given
+  /// child Widget will contain all of the [children].
+  final _ToolbarBuilder toolbarBuilder;
+
+  // Build the default Android Material text selection menu toolbar.
+  static Widget _defaultToolbarBuilder(BuildContext context, Widget child) {
+    return _TextSelectionToolbarContainer(
+      child: child,
+    );
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final double paddingTop = MediaQuery.of(context).padding.top
+        + _kToolbarScreenPadding;
+    final double availableHeight = anchorAbove.dy
+      - paddingTop;
+    final bool fitsAbove = _kToolbarHeight <= availableHeight;
+
+    return Stack(
+      children: <Widget>[
+        CustomSingleChildLayout(
+          delegate: _TextSelectionToolbarLayoutDelegate(
+            fitsAbove ? anchorAbove : anchorBelow,
+            paddingTop,
+            fitsAbove,
+          ),
+          child: _TextSelectionToolbarOverflowable(
+            isAbove: fitsAbove,
+            toolbarBuilder: toolbarBuilder,
+            children: children,
+          ),
+        ),
+      ],
+    );
+  }
+}
+
+// Positions the toolbar at the given anchor, ensuring that it remains on
+// screen.
+class _TextSelectionToolbarLayoutDelegate extends SingleChildLayoutDelegate {
+  _TextSelectionToolbarLayoutDelegate(
+    this.anchor,
+    this.upperBounds,
+    this.fitsAbove,
+  );
+
+  /// Anchor position of the toolbar in global coordinates.
+  final Offset anchor;
+
+  /// The upper-most valid y value for the anchor.
+  final double upperBounds;
+
+  /// Whether the closed toolbar fits above the anchor position.
+  ///
+  /// If the closed toolbar doesn't fit, then the menu is rendered below the
+  /// anchor position. It should never happen that the toolbar extends below the
+  /// padded bottom of the screen.
+  final bool fitsAbove;
+
+  // Return the value that centers width as closely as possible to position
+  // while fitting inside of min and max.
+  static double _centerOn(double position, double width, double min, double max) {
```

Comment:
Thanks for pointing this out, I tried playing around with it and I found that my constraints were not including the padding.

After fixing that, yes, the constraints will limit the child so that it doesn't exceed `max-min`.  If the developer does something tricky to exceed the width constraints, this logic will position it at the furthest left possible and let it overflow on the right.